go/pkg/cas/upload.go

```diff
@@ -970,3 +1105,79 @@ func marshalledRequestSize(d *repb.Digest) int64 {
 	}
 	return marshalledFieldSize(reqSize)
 }
+
+type missingDir struct {
+	relPath  string
+	item     *uploadItem
+	digested *digested
+}
+
+// missingDirs returns missing directories on the node path from the root to
+// each of the path specified in the tree.
+// For example, if the tree contains only "foo/bar" and "foo/baz", then
+// missingDirs returns "foo" and the root ".".
+//
+// All tree keys must be clean relative paths.
+func missingDirs(tree map[string]*digested, absRoot string) []missingDir {
+	// Establish parent->child edges.
+	children := map[string]map[string]struct{}{}
+	for relPath := range tree {
+		for relPath != "." {
+			parent := dirNameRelFast(relPath)
+			if childSet := children[parent]; childSet != nil {
+				childSet[relPath] = struct{}{}
+			} else {
+				children[parent] = map[string]struct{}{relPath: {}}
+			}
+			relPath = parent
+		}
+	}
+
+	// Add the missing directories by traversing in post-order.
+	var missing []missingDir
+	var dfs func(relPath string) proto.Message
+	dfs = func(relPath string) proto.Message {
+		if dig := tree[relPath]; dig != nil {
```

Comment:
Better to check like
```suggestion
		if dig, ok := tree[relPath]; ok {
```

Or is there a case where `tree[relPath] == nil` becomes true?

Reply:
no, there is no such case.